src/test/java/com/alphawallet/attestation/PublicAttestationTest.java

```diff
@@ -55,6 +62,62 @@ public void testNFTAttestation() throws Exception
+        assertArrayEquals(signedNFTAttestation2.getDerEncoding(), nftAttestation.getDerEncoding());
+    }
+
+    @Test
+    public void checkNFTSmartContract() throws Exception
+    {
+        ERC721Token myNFT = new ERC721Token("0xa567f5A165545Fa2639bBdA79991F105EADF8522", "25");
+        NFTAttestation nftAtt = new NFTAttestation(attestation, myNFT);
+        //construct SignedNFTAttestation using subject key
+        nftAttestation = new SignedNFTAttestation(nftAtt, subjectKeys); // <-- signing step, NFT attestation is signed by owner of identity, referenced below
+
+        //Extract the Ethereum signature
+        byte[] sig = nftAttestation.getSignature();
+
+        //generate NFTAttestation from the NFTAttestation bytes
+        NFTAttestation nftAttestation2 = new NFTAttestation(nftAtt.getDerEncoding(), attestorKeys.getPublic());
+
+        //check recovered signed attestation within the wrapping
+        assertTrue(nftAttestation2.verify());
+
+        //Generate SignedNFTAttestation using the reconstructed NFTAttestation and the extracted Ethereum signature
+        SignedNFTAttestation signedNFTAttestation2 = new SignedNFTAttestation(nftAttestation2, subjectKeys.getPublic(), sig);
+        assertTrue(signedNFTAttestation2.checkValidity());
+        assertTrue(nftAttestation.checkValidity());
+        assertArrayEquals(signedNFTAttestation2.getUnsignedAttestation().getDerEncoding(), nftAtt.getDerEncoding());
+        assertArrayEquals(signedNFTAttestation2.getDerEncoding(), nftAttestation.getDerEncoding());
+
+        //now check the attestation smart contract decodes this correctly
+        //first verify the public attestation
+
+        //Note that we use 'issuerKeys' address here as an unrelated address.
+        // This is to test the contract correctly checks that the owner (subjectKeys) has correctly signed the wrapping NFT attestation,
+        // which happens in the 'signing step' referenced above.
+        // This is so a transaction need not be sent by the 'subjectKeys' account. If it is called by
+        // 'subjectKeys' then the verification step is skipped since we have 'subjectKeys' signature from the ethereum transaction
+        AttestationReturn atr = contract.callVeryifyNFTAttestation(nftAttestation.getDerEncoding(), SignatureUtility.addressFromKey(issuerKeys.getPublic()));
+        //check our return
+        assertEquals(atr.identity, attestationIdentifier);
+        assertEquals(atr.ownerAddress.toLowerCase(), SignatureUtility.addressFromKey(subjectKeys.getPublic()).toLowerCase());
+        assertTrue(atr.isValid);
+        ERC721TokenEth token = atr.ercToken[0];
+        assertEquals(token.address.toString().toLowerCase(), myNFT.address.toLowerCase());
+        assertEquals(myNFT.tokenId, token.tokenId.getValue());
+
+        //TODO: make a more comprehensive negative test, involving bad attestation subject address etc.
+        //TODO: can do this if we remove the 'isValid' check in the constructor
```

Comment:
I think it would definitely make things easier in regards to such negative tests